From c052fc0e67fb8143436de58df8064859d0ca9d2c Mon Sep 17 00:00:00 2001
From: Tobias <ruck.tobias@gmail.com>
Date: Thu, 21 Nov 2024 19:00:49 +0100
Subject: [PATCH] [DogeCash] Speed up header presync/sync using a faster
 GetPowHash (#60)

Currently, GetPowHash allocates, which slows down the hash calculation. This is unnecessary and removing this speeds up the header presync/sync.

(cherry picked from commit a907d5e0ee925f241f53590e712427e6ffa902b3)

diff --git a/src/primitives/baseheader.cpp b/src/primitives/baseheader.cpp
index e5b00ce0086541f3b80e62f2fb418a89c922d4dd..ecaa837e1ae95274be38f732632457d244aa6779 100644
--- a/src/primitives/baseheader.cpp
+++ b/src/primitives/baseheader.cpp
@@ -14,9 +14,36 @@ BlockHash CBaseBlockHeader::GetHash() const {
 }
 
 BlockHash CBaseBlockHeader::GetPowHash() const {
-    CDataStream ss(SER_NETWORK, PROTOCOL_VERSION);
-    ss << *this;
+    // TODO: Dedup serialization, e.g. using a SpanWriter
+    uint8_t bytes[80];
+    size_t idx = 0;
+    uint32_t version = this->nVersion;
+    for (size_t i = 0; i < 4; ++i) {
+        bytes[idx++] = version & 0xff;
+        version >>= 8;
+    }
+    for (uint8_t byte : this->hashPrevBlock) {
+        bytes[idx++] = byte;
+    }
+    for (uint8_t byte : this->hashMerkleRoot) {
+        bytes[idx++] = byte;
+    }
+    uint32_t time = this->nTime;
+    for (size_t i = 0; i < 4; ++i) {
+        bytes[idx++] = time & 0xff;
+        time >>= 8;
+    }
+    uint32_t bits = this->nBits;
+    for (size_t i = 0; i < 4; ++i) {
+        bytes[idx++] = bits & 0xff;
+        bits >>= 8;
+    }
+    uint32_t nonce = this->nNonce;
+    for (size_t i = 0; i < 4; ++i) {
+        bytes[idx++] = nonce & 0xff;
+        nonce >>= 8;
+    }
     uint256 hash;
-    scrypt_1024_1_1_256((const uint8_t *)ss.data(), hash.data());
+    scrypt_1024_1_1_256(bytes, hash.data());
     return BlockHash(hash);
 }
